core/ctree/cnode.cpp

```diff
@@ -0,0 +1,420 @@
+        for(int i = 0; i < this->root_num; ++i){
+            this->roots[i].expand(0, 0, i, value_prefixs[i], policies[i]);
+            this->roots[i].add_exploration_noise(root_exploration_fraction, noises[i]);
+
+            this->roots[i].visit_count += 1;
+        }
+    }
+
+    void CRoots::prepare_no_noise(const std::vector<float> &value_prefixs, const std::vector<std::vector<float>> &policies){
+        for(int i = 0; i < this->root_num; ++i){
+            this->roots[i].expand(0, 0, i, value_prefixs[i], policies[i]);
+
+            this->roots[i].visit_count += 1;
+        }
+    }
+
+    void CRoots::clear(){
+        this->node_pools.clear();
+        this->roots.clear();
+    }
+
+    std::vector<std::vector<int>> CRoots::get_trajectories(){
+        std::vector<std::vector<int>> trajs;
+        trajs.reserve(this->root_num);
+
+        for(int i = 0; i < this->root_num; ++i){
+            trajs.push_back(this->roots[i].get_trajectory());
+        }
+        return trajs;
+    }
+
+    std::vector<std::vector<int>> CRoots::get_distributions(){
+        std::vector<std::vector<int>> distributions;
+        distributions.reserve(this->root_num);
+
+        for(int i = 0; i < this->root_num; ++i){
+            distributions.push_back(this->roots[i].get_children_distribution());
+        }
+        return distributions;
+    }
+
+    std::vector<float> CRoots::get_values(){
+        std::vector<float> values;
+        for(int i = 0; i < this->root_num; ++i){
+            values.push_back(this->roots[i].value());
+        }
+        return values;
+    }
+
+    //*********************************************************
+
+    void update_tree_q(CNode* root, tools::CMinMaxStats &min_max_stats, float discount){
+        std::stack<CNode*> node_stack;
+        node_stack.push(root);
+        float parent_value_prefix = 0.0;
+        int is_reset = 0;
+        while(node_stack.size() > 0){
+            CNode* node = node_stack.top();
+            node_stack.pop();
+
+            if(node != root){
+                float true_reward = node->value_prefix - parent_value_prefix;
+                if(is_reset == 1){
+                    true_reward = node->value_prefix;
+                }
+                float qsa = true_reward + discount * node->value();
+                min_max_stats.update(qsa);
+            }
+
+            for(int a = 0; a < node->action_num; ++a){
+                CNode* child = node->get_child(a);
+                if(child->expanded()){
+                    node_stack.push(child);
+                }
+            }
+
+            parent_value_prefix = node->value_prefix;
+            is_reset = node->is_reset;
+        }
+    }
+
+    void cback_propagate(std::vector<CNode*> &search_path, tools::CMinMaxStats &min_max_stats, int to_play, float value, float discount){
+        float bootstrap_value = value;
```

Comment:
Be a little careful when it comes to the value and reward. I would print the output of the prediction model to make sure that these are 1 value. I know the value head and reward head are 601 outputs that are then encoded into 1 but I would double check where that happens.